lib/engine/game/g_1832/step/buy_sell_par_shares.rb

```diff
@@ -10,6 +10,25 @@ class BuySellParShares < G1870::Step::BuySellParShares
           def visible_corporations
             @game.sorted_corporations.reject { |item| item.type == :system unless item.floated? }
           end
+
+          def can_exchange_any?(entity)
+            return false unless entity.player?
+
+            entity.companies.each do |company|
+              next unless company.id == 'P4'
+
+              # Only true if there are shares available from an ipoed corp that has not yet operated
+              corps = @game.corporations.select(&:ipoed).reject do |c|
+                c.operating_history.size.positive?
+              end
+
+              corps.each do |c|
+                return true unless c.num_ipo_shares.empty
+              end
+            end
+
+            false
+          end
```

Comment:
`num_shares_of` returns a number, so testing `empty?` (which needs a question mark) isn't appropriate. But you should use `empty?` test on the `operating_history` hash, instead of testing `size.positive?`.

You could also combine the two iterators over corporations to make the code clearer. And I think it would be best if the corporation iterator was moved out of the companies iterator, as they are not related.
```suggestion
            return false if entity.companies.none? { |company| company.id == 'P4' }
            
            # Can only exchange if there are shares available from an ipoed corp that has not yet operated
            @game.corporations.any? do |corp|
              corp.ipoed && !corp.operating_history.empty? && corp.num_ipo_shares.positive?
            end
          end
```